lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -0,0 +1,202 @@
+module Cardano.Wallet.Shelley.Pools where
+
+import Prelude
+
+import Cardano.Wallet.Api.Server
+    ( LiftHandler (..), apiError )
+import Cardano.Wallet.Api.Types
+    ( ApiErrorCode (..), ApiT (..) )
+import Cardano.Wallet.Network
+    ( NetworkLayer (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..), GenesisParameters (..), PoolId )
+import Cardano.Wallet.Shelley.Compatibility
+    ( Shelley
+    , ShelleyBlock
+    , fromPoolDistr
+    , fromRewards
+    , optimumNumberOfPools
+    , toPoint
+    , toShelleyCoin
+    )
+import Cardano.Wallet.Shelley.Network
+    ( pattern Cursor )
+import Cardano.Wallet.Unsafe
+    ( unsafeMkPercentage, unsafeRunExceptT )
+import Control.Monad.Class.MonadSTM
+    ( MonadSTM, TQueue )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), runExceptT, withExceptT )
+import Data.Map
+    ( Map )
+import Data.Map.Merge.Strict
+    ( dropMissing, traverseMissing, zipWithMatched )
+import Data.Ord
+    ( Down (..) )
+import Data.Quantity
+    ( Percentage (..), Quantity (..) )
+import Data.Sort
+    ( sortOn )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Ouroboros.Network.Block
+    ( Point )
+import Ouroboros.Network.Client.Wallet
+    ( LocalStateQueryCmd (..), send )
+import Servant
+    ( Handler, err500 )
+
+import qualified Cardano.Wallet.Api.Types as Api
+import qualified Data.Map.Merge.Strict as Map
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import qualified Ouroboros.Consensus.Shelley.Ledger as OC
+
+-- | Stake Pool Data fields fetched from the node via LSQ
+data PoolLsqMetrics = PoolLsqMetrics
+    { nonMyopicMemberRewards :: Quantity "lovelace" Word64
+    , stake :: Percentage
+    , saturation :: Double
+    } deriving (Eq, Show, Generic)
+
+
+data ErrFetchMetrics = ErrFetchMetrics
+  deriving Show
+
+askNode
+    :: MonadSTM m
+    => TQueue m (LocalStateQueryCmd ShelleyBlock m)
+    -> Point ShelleyBlock
+    -> Coin
+    -> ExceptT ErrFetchMetrics m (Map PoolId PoolLsqMetrics)
+askNode queue pt coin = do
+    stakeMap <- fromPoolDistr <$> handleQueryFailure
+        (queue `send` CmdQueryLocalState pt OC.GetStakeDistribution)
+    let toStake = Set.singleton $ Left $ toShelleyCoin coin
+    rewardMap <- fromRewards <$> handleQueryFailure
+        (queue `send` CmdQueryLocalState pt (OC.GetNonMyopicMemberRewards toStake))
+    pparams <- handleQueryFailure
+        (queue `send` CmdQueryLocalState pt OC.GetCurrentPParams)
+
+    return $ combine
+        (optimumNumberOfPools pparams)
+        stakeMap
+        rewardMap
+  where
+    handleQueryFailure = withExceptT (const ErrFetchMetrics) . ExceptT
+
+    combine
+        :: Int -- ^ Desired number of pools
+        -> Map PoolId Percentage
+        -> Map PoolId (Quantity "lovelace" Word64)
+        -> Map PoolId PoolLsqMetrics
+    combine nOpt =
+        Map.merge stakeButNoRew rewardsButNoStake bothPresent
```

Comment:
How is it actually possible with the LSQ to have cases where a poolId wouldn't be in both maps? From the excerpt above, I get that these are two separate requests, but the local state query allows for "locking" a state and making several queries on that locked state if I recall correctly? So in principle, that's sort of a guarantee that the list of pools fetched from the stake distribution is exactly the same as the list of pools get from the member rewards.

Reply:
The poolId may not be in the rewards map (clarified comment in cacff5590ed5f77ba6aa0d5e6420e968733f1f99)

As for the `rewardsButNoStake` case: Yes, I suspect that is impossible. (Maybe @rvl knows for sure though?)

I guess we should have some `InconcistentMetrics` similar to what we had for jormungandr, instead of the current silent `dropMissing`.

We will have more problems like this when we start folding over pool registration certificates on the chain.

Reply:
> Yes, I suspect that is impossible
But I guess regardless of locking or not, we are making the queries with the same specific `Point`, so they should all be consistent if they succeed 🤔

Reply:
> I guess we should have some InconcistentMetrics similar to what we had for jormungandr, instead of the current silent dropMissing.

Not really though. In the case of Jörmungandr, we knew we could end up with inconsistent metrics because of the concurrent nature of our logic + the stateless API from Jörmungandr. So it was kind of _expected_. With cardano-node here, this is really something we do not expect, and therefore,not something we're willing to push onto the caller of this function.